application/common/models/Method.php

```diff
@@ -61,4 +93,159 @@ public function getRawPhoneNumber()
+        }
+
+        if ( ! $method->save()) {
+            $log['status'] = 'failed';
+            $log['error'] = Json::encode($method->getFirstErrors());
+            \Yii::error($log);
+
+            throw new \Exception('Unable to add new method', 1461375342);
+        }
+
+        /*
+         * Method saved, send verification
+         */
+        $method->sendVerification();
+
+        $log['status'] = 'success';
+        \Yii::warning($log);
+
+        return $method;
+    }
+
+    /**
+     * Send verification to either email or phone based on $this->type
+     * @throws \Exception
+     */
+    public function sendVerification()
+    {
+        if ($this->type == self::TYPE_EMAIL) {
+            $this->sendVerificationEmail();
+        } elseif ($this->type == self::TYPE_PHONE) {
+            $this->verification_code = $this->sendVerificationPhone();
+            if ( ! $this->save()) {
+                throw new \Exception('Unable to save method after sending phone verification', 1461441850);
+            }
+        } else {
+            throw new \Exception('Invalid method type', 1461432437);
+        }
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function sendVerificationEmail()
+    {
+        Verification::sendEmail(
+            $this->value,
+            'Verification required - New account recovery method added',
+            '@common/mail/method/verify',
+            $this->verification_code,
+            $this->user,
+            null,
+            $this->user->getId(),
+            'New email method',
+            'A new email method has been added and verification sent to ' . $this->getMaskedValue()
+        );
+    }
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    public function sendVerificationPhone()
+    {
+        return Verification::sendPhone(
+            $this->getRawPhoneNumber(),
+            $this->verification_code,
+            $this->user->getId(),
+            'New phone method',
+            'A new phone method has been added and verification sent to ' . $this->getMaskedValue()
+        );
+    }
+
+    /**
+     * @param string $userSubmitted
+     * @return bool
+     */
+    public function isUserProvidedCodeCorrect($userSubmitted)
+    {
+        if ($this->type === self::TYPE_EMAIL) {
+            return Verification::isEmailCodeValid($this->verification_code, $userSubmitted);
+        } elseif ($this->type === self::TYPE_PHONE) {
+            return Verification::isPhoneCodeValid($this->verification_code, $userSubmitted);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Validate user submitted code and update record to be verified if valid
+     * @param string $userSubmitted
+     * @throws \Exception
+     */
+    public function validateAndSetAsVerified($userSubmitted)
+    {
+        if ( ! $this->isUserProvidedCodeCorrect($userSubmitted)) {
+            $this->verification_attempts++;
+            $this->save();
```

Comment:
We're already throwing an exception right after it so I'm not sure what we'd do if it failed anyway, ideas?